<launch>
    <arg name="device" default="1"
         doc="target device to execute inference. [0:CPU]|1:GPU|2:VPU|3:FPGA"/>

    <!-- Load hand geometry parameters -->
    <include file="$(find gpd_ecsam)/launch/j2n6s200_hand_geometry.launch.xml">
        <arg name="node" value="detect_grasps"/>
    </include>

    <!-- Load classifier parameters -->
    <include file="$(find gpd_ecsam)/launch/caffe/j2n6s200_classifier_15_channels.launch.xml">
        <arg name="node" value="detect_grasps"/>
    </include>

    <node name="detect_grasps" pkg="gpd" type="detect_grasps" output="screen">

        <param name="device" type="int" value="$(arg device)"/>

        <!-- If sequential importance sampling is used (default: false) -->
        <param name="use_importance_sampling" value="false"/>

        <!-- What type of point cloud is used and what ROS topic it comes from -->
        <param name="cloud_type" value="0"/> <!-- 0: PointCloud2, 1: CloudSized, 2: CloudIndexed, 3: CloudSamples -->
        <param name="cloud_topic" value="/ecsam/correlated_object_point_cloud"/>

        <!-- (optional) The ROS topic that the samples come from (default: an empty string) -->
        <param name="samples_topic" value=""/>

        <!-- Plotting parameters -->
        <param name="plot_normals" value="false"/>
        <param name="plot_samples" value="false"/>
        <param name="plot_candidates" value="false"/>
        <param name="plot_filtered_grasps" value="false"/>
        <param name="plot_valid_grasps" value="false"/>
        <param name="plot_clusters" value="false"/>
        <param name="plot_selected_grasps" value="false"/>
        <param name="rviz_topic" value="grasps_rviz"/>

        <!-- Preprocessing of point cloud -->
        <param name="voxelize" value="false"/>
        <param name="remove_outliers" value="false"/>
        <rosparam param="workspace">[-1.0, 1.0, -1.0, 1.0, -1.0, 1.0]</rosparam>
        <rosparam param="camera_position">[0.019, -0.494, 0.2415]</rosparam>

        <!-- General parameters -->
        <param name="num_samples" value="500"/>
        <param name="num_threads" value="4"/>
        <!--    <param name="gaussian_table_size" value="10000000" />-->

        <!-- Parameters for local grasp candidate search -->
        <param name="nn_radius" value="0.01"/>
        <param name="num_orientations" value="8"/> <!-- Number of orientations to consider -->

        <!-- Filtering of grasp candidates -->
        <param name="filter_grasps" value="true"/> <!-- on workspace and robot hand aperture -->
        <rosparam param="workspace_grasps">[-1.0, 1.0, -1.0, 1.0, -1.0, 1.0]</rosparam>
        <param name="filter_half_antipodal" value="false"/> <!-- on half antipodal -->

        <!-- Grasp image creation -->
        <param name="create_image_batches" value="false"/> <!-- creates grasp images in batches (less memory usage) -->
        <param name="remove_plane_before_image_calculation"
               value="false"/> <!-- removes table plane from point cloud to speed up shadow computations -->

        <!-- Clustering of grasps -->
        <param name="min_inliers" value="1"/>

        <!-- Grasp selection -->
        <param name="min_score_diff" value="100"/>
        <param name="min_aperture" value="0.04"/>
        <param name="max_aperture" value="0.08"/>
        <!--    <param name="min_aperture" value="0.029" />-->
        <!--    <param name="max_aperture" value="0.072" />-->
        <param name="num_selected" value="500"/>

    </node>

</launch>
